### DISCUSSION TITLE
usage/engine/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9179] [Docs]
# usage/engine/

Discover how to customize and extend base Ultralytics YOLO Trainer engines. Support your custom model and dataloader by overriding built-in functions.

https://docs.ultralytics.com/usage/engine/

**Comment by UttamToni** [comment]
As mentioned in the advanced customization of ultralytics:

from ultralytics.models.yolo.detect import DetectionTrainer
trainer = DetectionTrainer(overrides={...})

i am confuced that which file should i place in the curly brackets of ```overrides={...}``` . Is the ```args.yaml``` file is allowed , which is created after the training step?
I have used the args.yaml file which is created after training step but got this error:
```
/usr/local/lib/python3.10/dist-packages/ultralytics/engine/trainer.py in __init__(self, cfg, overrides, _callbacks)
     96             overrides (dict, optional): Configuration overrides. Defaults to None.
     97         """
---> 98         self.args = get_cfg(cfg, overrides)
     99         self.check_resume(overrides)
    100         self.device = select_device(self.args.device, self.args.batch)

/usr/local/lib/python3.10/dist-packages/ultralytics/cfg/__init__.py in get_cfg(cfg, overrides)
    207         overrides = cfg2dict(overrides)
    208         if "save_dir" not in cfg:
--> 209             overrides.pop("save_dir", None)  # special override keys to ignore
    210         check_dict_alignment(cfg, overrides)
    211         cfg = {**cfg, **overrides}  # merge cfg and overrides dicts (prefer overrides)

TypeError: set.pop() takes no arguments (2 given)

```

**Reply by glenn-jocher** [reply]
Hello! It seems like there’s been a little confusion regarding the `overrides={...}` usage, and I appreciate you reaching out for clarification. 🙌

To directly answer your question, the `overrides={...}` parameter is designed to accept a dictionary of configuration overrides, not a file. It's intended to programmatically override specific configuration entries. For example:

```python
from ultralytics.models.yolo.detect import DetectionTrainer

# Example override dictionary
overrides = {
    "batch_size": 16,
    "img_size": 640,
}

trainer = DetectionTrainer(overrides=overrides)
```

If you're looking to use the configurations from an `args.yaml` file created after training, you'd first need to read this YAML file into a Python dictionary, which you could then pass to `overrides`. Here's a quick way to do it:

```python
import yaml

# Load your args.yaml content into a dictionary
with open('path/to/your/args.yaml', 'r') as file:
    yaml_content = yaml.safe_load(file)

# Now, pass the loaded dictionary to overrides
trainer = DetectionTrainer(overrides=yaml_content)
```

Regarding the error you're encountering, it appears there's a misunderstanding in how the `overrides` dictionary is being used or possibly an issue with the implementation. If following the above pattern doesn't resolve your issue, could you please share more details about what you're trying to achieve with the `overrides`? This way, I or someone from the community can offer more targeted advice. 😊

Thank you for your patience, and looking forward to helping you get this sorted!

**Comment by ternaus** [comment]
Is it possible to define custom DataLoader or Augmentation pipeline in DetectionTrainer?

**Reply by pderrenger** [reply]
Yes, it's definitely possible to define custom DataLoader and Augmentation pipelines in DetectionTrainer! 🚀 To do so, you'd primarily override the `get_dataloader()` method in a subclass of `DetectionTrainer`. Here's a brief example on how you might go about customizing your DataLoader:

```python
from ultralytics.models.yolo.detect import DetectionTrainer

class CustomTrainer(DetectionTrainer):
    def get_dataloader(self, ...):
        # Your custom DataLoader logic here
        ...
```

For augmentations, you can integrate them within your custom DataLoader or modify datasets accordingly. If you need further customization or to include a custom Augmentation pipeline directly tied with the models' training loop, you could adjust or extend methods in the model itself or within the `get_dataloader` method to apply augmentations.

Don't forget to explore our [Advanced Customization](https://docs.ultralytics.com/usage/engine/) section and [BaseTrainer Reference](https://docs.ultralytics.com/usage/engine/) for more details on customizing training routines, models, and DataLoaders. 😄 Happy coding!

**Comment by ablyzniuk** [comment]
How to set a custom loss function?

**Reply by ablyzniuk** [reply]
Sorry, I missed it above. I saw it))) Delete this comment.

**Reply by glenn-jocher** [reply]
@ablyzniuk to set a custom loss function, you can override the `init_criterion` method in your custom model class, as shown in the [Advanced Customization documentation](https://docs.ultralytics.com/usage/engine/). This method allows you to define and initialize your loss function according to your requirements. Let me know if you need an example!

**Reply by ablyzniuk** [reply]
@glen-jocher thank you) I will be glad and thankfull if you will give a detailed example of training default Ultralytics Yolo(5-11 versions) with custom trainer e.t.c. I think there must be much easier and convinient way opposite what I did (I edited source code and overrides DetectionTrainer methods after)

**Reply by glenn-jocher** [reply]
For custom training without modifying source code, subclass `DetectionTrainer` and override methods like this:

```python
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.nn.tasks import DetectionModel

class CustomModel(DetectionModel):
    def init_criterion(self):
        # Your custom loss initialization here
        pass

class CustomTrainer(DetectionTrainer):
    def get_model(self, cfg, weights):
        return CustomModel(cfg)  # Replace with your model config

trainer = CustomTrainer(overrides={'model': 'yolov8n.yaml', 'data': 'coco128.yaml', 'epochs': 100})
trainer.train()
```

This maintains upgrade compatibility while allowing customization. The [Advanced Customization Guide](https://docs.ultralytics.com/usage/engine/) shows additional methods you can override.
